### ISSUE TITLE
GCS Sensor for matching glob

### ISSUE THREAD

**Issue by A-Costa** [apache/airflow#32650]
### Description

Since the new version of [`GCSHook.list`](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/_api/airflow/providers/google/cloud/hooks/gcs/index.html#airflow.providers.google.cloud.hooks.gcs.GCSHook.list) implements the `match_glob` parameter, it would be possible to implement a new GCS sensor that checks if a file that matches a given glob exists in a GCS bucket.

The new sensor could be called `GCSObjectsMatchingGlobExistenceSensor` and would have a similar signature to  [`GCSObjectsWithPrefixExistenceSensor`](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/_api/airflow/providers/google/cloud/sensors/gcs/index.html#airflow.providers.google.cloud.sensors.gcs.GCSObjectsWithPrefixExistenceSensor), changing in the `__init__` function the `prefix` parameter with a `glob` parameter.

### Use case/motivation

This will be a new sensor, useful to check if inside a Google Cloud Storage bucket a file that matches a certain blob is available. 

For example:

```python
GCSObjectsMatchingGlobExistenceSensor(
    task_id="gcs_file_sensor",
    bucket="test_bucket",
    glob="*.csv"
)
```

will find all csv files in the bucket `test_bucket`.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by A-Costa** [comment]
I am currently working on a PR. Starting with [GCSObjectsWithPrefixExistenceSensor](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/_api/airflow/providers/google/cloud/sensors/gcs/index.html#airflow.providers.google.cloud.sensors.gcs.GCSObjectsWithPrefixExistenceSensor) as a base, i have changed the `poke` function to use the `match_glob` parameter when doing `hook.list`, however i have now realised that i might need to define a new trigger.

**Comment by eladkal** [comment]
I suggest to double check why we have `GCSObjectsWithPrefixExistenceSensor` & `GCSObjectExistenceSensor` to begin with?
`prefix` should be a parameter of  `GCSObjectExistenceSensor` or even consider generalizing further like [wildcard_match](https://github.com/apache/airflow/blob/c99c03135ca7e7c41e1c6d338de9e41422ea84f0/airflow/providers/amazon/aws/sensors/s3.py#L57) of `S3KeySensor` that can handle more than just prefix.

**Comment by A-Costa** [comment]
I agree! It would make a lot of sense to generalise `GCSObjectExistenceSensor` and add both `prefix` and `glob` there as parameters

**Comment by pankajkoti** [comment]
@A-Costa assigned it to you assuming you're still on it

**Comment by A-Costa** [comment]
Might need some help from someone more experienced, but I can work on it!

**Comment by shahar1** [comment]
@A-Costa Hey, if you're still working on it - I'd be happy to assist (disclosure: I introduced the `match_glob` to the GCS hook).
As Elad suggested - I think that it makes sense to first deprecate `GCSObjectsWithPrefixExistenceSensor` in favor of `GCSObjectExistenceSensor`, as the former is a private case of the latter. Then, introducing the `match_glob` to the `GCSObjectExistenceSensor` sensor shouldn't be too much work :)
When you have time - try to come up with an initial PR and I could give you some feedback.

**Comment by A-Costa** [comment]
Hi @shahar1, sounds good! I'll try to come up with a PR by the end of this week, thanks for your support

**Comment by A-Costa** [comment]
Hi @shahar1 i've published a very rudimental draft PR #34137 

The first issue i'm encountering is the following: while it was easy to adapt the `poke` method to use the `match_glob` parameter, it is not easy to adjust the `execute` method, in particular when the sensor is set with `deferrable=True`.
The problem here is that the sensor is using the `GCSBlobTrigger`, which in turn uses the [`Bucket`](https://github.com/talkiq/gcloud-aio/blob/b170cf916f0ad52357bd3457ce5b905dd9170132/storage/gcloud/aio/storage/bucket.py#L30) class from the [gcloud-aio](https://github.com/talkiq/gcloud-aio) library.

The `GCSBlobTrigger` is using the method `blob_exists` which only works with the exact filename.

https://github.com/apache/airflow/blob/b9acffa81bf61dcf0c5553942c52629c7f75ebe2/airflow/providers/google/cloud/triggers/gcs.py#L101

Another method called [`list_blobs`](https://github.com/talkiq/gcloud-aio/blob/b170cf916f0ad52357bd3457ce5b905dd9170132/storage/gcloud/aio/storage/bucket.py#L64) is available in the class and is in fact used by `GCSPrefixBlobTrigger`. The problem is that `list_blob` only implements the `prefix` parameter and not a glob matching one.

I'm now gonna open an issue on the `gcloud-aio` repository and see if they are willing to add this functionality, otherwise i guess the only viable approach would be to implement it ourselves. 

Basically it's the same issue that you had to solve implementing [`_list_blobs_with_match_glob`](https://github.com/apache/airflow/blob/b9acffa81bf61dcf0c5553942c52629c7f75ebe2/airflow/providers/google/cloud/hooks/gcs.py#L847) but for the async version of it.

**Comment by A-Costa** [comment]
Ok so the change in the [gcloud-aio](https://github.com/talkiq/gcloud-aio) library seemed much simpler than i thought. I've opened a [PR in their repository](https://github.com/talkiq/gcloud-aio/pull/634), tomorrow i'll draft a solution here that uses `gcloud-aio` with my proposed change

**Comment by eladkal** [comment]
solved in https://github.com/apache/airflow/pull/34137
